File: moderate/chain_inspection.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

func main() {
	var found bool
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	scanner := bufio.NewScanner(data)
	for scanner.Scan() {
		var been []string
		s := strings.Split(scanner.Text(), ";")
		m, b := make(map[string]string, len(s)), "BEGIN"
		for _, i := range s {
			t := strings.Split(i, "-")
			if t[0] == t[1] {
				goto invalid
			}
			m[t[0]] = t[1]
		}
		for _ = range s {
			if b, found = m[b]; !found {
				goto invalid
			}
			for _, j := range been {
				if b == j {
					goto invalid
				}
			}
			been = append(been, b)
		}
		if b == "END" {
			fmt.Println("GOOD")
			continue
		}
	invalid:
		fmt.Println("BAD")
	}
}

File: moderate/chain_inspection.rb
File.open(ARGV[0]).each_line do |line|
  b, s, m, been = "BEGIN", line.chomp.split(";"), Hash.new, Hash.new
  s.each do |x|
    t = x.split("-")
    m[t[0]] = t[1]
  end
  s.length.times do
    if !(m.has_key?(b)) || (been.has_key?(m[b]))
      b = nil
      break
    end
    b = m[b]
    been[b] = true
  end
  puts b == "END" ? "GOOD" : "BAD"
end
